hub/util/cache_chain.py

```diff
@@ -23,15 +25,46 @@ def get_cache_chain(provider_list: List[StorageProvider], size_list: List[int]):
         ProviderListEmptyError: If the provider list is empty.
         ProviderSizeListMismatch: If the len(size_list) + 1 != len(provider_list)
     """
-    if not provider_list:
+    if not storage_list:
         raise ProviderListEmptyError
-    if len(provider_list) <= 1:
-        return provider_list[0]
-    if len(size_list) + 1 != len(provider_list):
+    if len(storage_list) <= 1:
+        return storage_list[0]
+    if len(size_list) + 1 != len(storage_list):
         raise ProviderSizeListMismatch
-    provider_list.reverse()
-    size_list.reverse()
-    store = provider_list[0]
-    for size, cache in zip(size_list, provider_list[1:]):
+    store = storage_list[-1]
+    for size, cache in zip(reversed(size_list), reversed(storage_list[:-1])):
         store = LRUCache(cache, store, size)
     return store
+
+
+def generate_chain(
+    base_storage: StorageProvider,
+    memory_cache_size: int,
```

Comment:
This should be a list of cache sizes. whose length is less than 1 to Storage provider. We should not assume that the cache chain can only have max of three providers.

Reply:
@imshashank that is a separate function in the same file. This is a function only for the dataset, that can only have 2 caches